### PR TITLE
Refactor downloadBlob into downloadFile

### PR DESCRIPTION
**Pull request by wwlorey**
This adds support for downloading single files and fixes https://github.com/microsoft/vscode-azurestorage/issues/755

### CODE REVIEW

File: src/tree/blob/BlobTreeItem.ts [hunk 1]
```diff
@@ -77,44 +73,7 @@ export class BlobTreeItem extends AzureTreeItem<IStorageRoot> implements ICopyUr
...
-
-            const title: string = localize('downloadingTo', 'Downloading {0} to {1}...', this.blobName, uri.fsPath);
-            ext.outputChannel.appendLog(title);
-            await window.withProgress({ title, location: vscode.ProgressLocation.Notification }, async (notificationProgress, cancellationToken) => {
-                await azCopyTransfer(context, 'BlobLocal', src, dst, transferProgress, notificationProgress, cancellationToken);
-            });
-
-            ext.outputChannel.appendLog(`Successfully downloaded ${linkablePath}.`);
-        }
-    }
```
Comment (wwlorey) [inline src/tree/blob/BlobTreeItem.ts:115]: I moved everything from this function into downloadFile.ts with the added logic for file downloads

File: src/commands/downloadFile.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    const uri: Uri = await getUriForDownload(remoteFileName);
+    if (uri.scheme === 'file') {
+        const linkablePath: Uri = Uri.file(uri.fsPath); // Allows CTRL+Click in Output panel
+        const src: IRemoteSasLocation = createAzCopyRemoteLocation(treeItem, remoteFilePath);
+        const dst: ILocalLocation = createAzCopyLocalLocation(uri.path);
+        // tslint:disable-next-line: strict-boolean-expressions
+        const totalBytes: number = (await storageClient.getProperties()).contentLength || 1;
+        const transferProgress: TransferProgress = new TransferProgress(totalBytes, remoteFileName);
+        const title: string = localize('downloadingTo', 'Downloading {0} to {1}...', remoteFileName, uri.fsPath);
```
Comment (EricJizbaMSFT) [inline src/commands/downloadFile.ts:52]:
```suggestion
        const title: string = localize('downloadingTo', 'Downloading from "{0}" to "{1}"...', remoteFileName, uri.fsPath);
```

File: src/commands/downloadFile.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+async function getUriForDownload(resourceName: string): Promise<Uri> {
+    const extension: string = extname(resourceName);
+    const filters = { "All files": ['*'] };
+    if (extension) {
+        // This is needed to ensure the file extension is added in the Save dialog, since the filename will be displayed without it by default on Windows
+        filters[`*${extension}`] = [extension];
+    }
+    const options: SaveDialogOptions = {
+        saveLabel: "Download",
```
Comment (EricJizbaMSFT) [inline src/commands/downloadFile.ts:70]:
```suggestion
        saveLabel: localize('download', 'Download'),
```

**Review by EricJizbaMSFT — APPROVED** [review]
I'm pretty sure my comments would've applied before you made any changes, but this is what you get for copying and changing code in the same PR 😝 If that is the case, I'm fine if you ignore my comments for the sake of time/risk

File: src/commands/downloadFile.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        storageClient = createBlockBlobClient(treeItem.root, treeItem.container.name, treeItem.blobPath);
+    } else {
+        remoteFileName = treeItem.fileName;
+        remoteFilePath = join(treeItem.directoryPath, treeItem.fileName);
+        fromTo = 'FileLocal';
+        storageClient = createFileClient(treeItem.root, treeItem.shareName, treeItem.directoryPath, treeItem.fileName);
+    }
+
+    const uri: Uri = await getUriForDownload(remoteFileName);
+    if (uri.scheme === 'file') {
```
[the remark this answers is not in the document]
Comment (wwlorey) [reply src/commands/downloadFile.ts:45]: I added an error if the scheme isn't "file"

File: src/commands/downloadFile.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        ext.outputChannel.appendLog(`Successfully downloaded ${linkablePath}.`);
+    }
+}
+
+async function getUriForDownload(resourceName: string): Promise<Uri> {
+    const extension: string = extname(resourceName);
+    const filters = { "All files": ['*'] };
+    if (extension) {
+        // This is needed to ensure the file extension is added in the Save dialog, since the filename will be displayed without it by default on Windows
+        filters[`*${extension}`] = [extension];
```
[the remark this answers is not in the document]
Comment (wwlorey) [reply src/commands/downloadFile.ts:69]: On windows, the default is the extension filter. If you choose the "All files" filter instead, you can use whatever extension you want as expected

**Pull request merged**